### PR TITLE
Add option to bypass HDF5_jll

### PR DESCRIPTION
**Pull request by jipolanco**
As discussed in #606, this is an initial proposal to provide a fallback mechanism for using a custom HDF5 build, instead of `HDF5_jll`.

This mechanism is enabled by setting the `JULIA_HDF5_LIBRARY_PATH` environment variable to the path of the HDF5 libraries at build time. By default, if this variable is not set, we use `HDF5_jll`.

The implementation is modelled after similar mechanisms used by FFTW.jl and MPI.jl.

### CODE REVIEW

**Comment by musm** [comment]
LGTM. Presumably this worked locally?

**Comment by jipolanco** [comment]
Yes, forgot to mention that. I tested it with HDF5 linked to MPICH, installed from the Ubuntu repos. MPI functionality works as well.

File: src/HDF5.jl [hunk 1]
```diff
@@ -2556,6 +2563,8 @@ function __init__()
 
     register_blosc()
 
+    check_deps()
```
Comment (musm) [inline src/HDF5.jl]: Maybe move this up to the start of the function? L2559

File: deps/build.jl [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                # Do not edit
+
+                function check_deps()
+                    if libhdf5_size != filesize(Libdl.dlpath(libhdf5))
+                        error("HDF5 library has changed, re-run Pkg.build(\\\"HDF5\\\")")
+                    end
+                end
+                """
+               )
+        println(io, :( const libhdf5 = $libhdf5 ))
```
Comment (musm) [inline deps/build.jl]:
```suggestion
        println(io, :(const libhdf5 = $libhdf5))
```

**Comment by musm** [comment]
Do you mind adding a very short statement about the env variable in the README as well?

File: deps/build.jl (hunk 2 above)
Comment (jipolanco) [reply deps/build.jl]: Updated this and the position of the `check_deps` call in 69e6c9d.

File: src/HDF5.jl [hunk 3]
```diff
@@ -26,6 +24,15 @@ export
     o_copy, o_delete, o_open, p_create,
     readmmap, @read, @write, root, set_dims!, t_create, t_commit
 
+# This is taken from FFTW.jl
```
Comment (musm) [inline src/HDF5.jl]: unecessary comment

Comment (musm) [inline src/HDF5.jl]:
```suggestion
```

**Comment by musm** [comment]
We'll need to update 
https://github.com/JuliaIO/HDF5.jl/blob/master/README.md#installation

with the new instructions here

**Comment by jipolanco** [comment]
Yes, I'm working on the instructions

**Comment by jipolanco** [comment]
I just pushed a proposal for the instructions, but feel free to make any changes

**Comment by musm** [comment]
Thanks a ton!

**Pull request merged**